In [1]:
import numpy as np
import pandas as pd
import random

import matplotlib.pyplot as plt
%matplotlib inline

plt.rcParams["figure.figsize"] = (12,6)

# NumPy

**Задача 1**\
Сгенерировать массив из случайных положительных чисел произвольной длины. Вывести элементы, которые больше среднего арифметического.

In [2]:
# Сгенерировать массив из случайных положительных чисел произвольной длины. Я выбрала числа от 1 до 100 и размерность массива - 13 элементов
arr = np.random.randint(1, 100, 13)
arr

array([13, 10, 58, 71, 86, 90, 18, 22, 65,  9, 79, 43, 11])

In [3]:
m = arr.mean()
m

44.23076923076923

In [4]:
# Вывести элементы, которые больше среднего арифметического.
listok = []
for i in arr:
    if i>m:
        listok.append(i)
print(listok)

[58, 71, 86, 90, 65, 79]


**Задача 2**\
Сгенерировать массив из 20 случайных элементов, элементы которого лежат в интервале [0, 10].
+ Определить количество элементов, отличающихся от минимального на 5.
+ Найти среднее арифметическое всех четных элементов массива, стоящих на нечетных местах.

In [5]:
# Сгенерировать массив из 20 случайных элементов, элементы которого лежат в интервале [0, 10].
arr_2 = np.random.randint(0, 11, 20)
arr_2

array([ 5,  9,  3,  0,  1, 10,  8,  9,  6,  2,  5,  6, 10,  8, 10,  9,  4,
        8, 10,  9])

In [6]:
arr_2.min()

0

In [7]:
# Определить количество элементов, отличающихся от минимального на 5.
listochek = []
for i in arr_2:
    if i - arr_2.min()==5:
        listochek.append(i)
print(len(listochek))

2


In [8]:
# начинаем с 0, 0 чётный
# Найти среднее арифметическое всех четных элементов массива, стоящих на нечетных местах.
li = []
for i in arr_2[1::2]:
    if i % 2 == 0:
        li.append(i)
        av = (sum(li)/len(li)).round(2)
print('Все чётные числа, стоящие на нечетных местах',li)
print('Среднее арифметическое',av)

Все чётные числа, стоящие на нечетных местах [0, 10, 2, 6, 8, 8]
Среднее арифметическое 5.67


**Задача 3**\
Сгенерировать матрицу A размером 5 на 5, элементы которой лежат в интервале [1, 10].
+ Возвести в квадрат максимальный элемент матрицы.
+ Все четные числа заменить на 0.

In [9]:
# Сгенерировать матрицу A размером 5 на 5, элементы которой лежат в интервале [1, 10].
my_arr = np.random.randint(1, 11, 25).reshape(5,5)
my_arr

array([[ 8,  2,  6,  3,  2],
       [ 3,  3,  4,  8,  8],
       [ 5,  8,  1,  2,  3],
       [ 1, 10,  3,  8,  4],
       [10,  7, 10,  2,  8]])

In [10]:
# Возвести в квадрат максимальный элемент матрицы.
el = my_arr.max()
el**2

100

In [11]:
# Все четные числа заменить на 0.
my_arr[my_arr % 2 == 0] = 0
my_arr

array([[0, 0, 0, 3, 0],
       [3, 3, 0, 0, 0],
       [5, 0, 1, 0, 3],
       [1, 0, 3, 0, 0],
       [0, 7, 0, 0, 0]])

**Задача 4**\
Сгенерировать матрицу B размером 5 на 5, элементы которой лежат в интервале [-5, 5]. Отрицательные элементы матрицы заменить на число отрицательных элементов в строке, в которой они находятся.

In [12]:
# Сгенерировать матрицу B размером 5 на 5, элементы которой лежат в интервале [-5, 5].
my_arr_2 = np.random.randint(-5, 6, 25).reshape(5,5)
my_arr_2

array([[ 3,  1,  2,  2, -5],
       [ 5, -4,  3,  5,  1],
       [ 2,  0, -4,  2,  5],
       [ 3,  2,  4,  0, -2],
       [-5,  0,  1,  4, -1]])

In [13]:
# Отрицательные элементы матрицы заменить на число отрицательных элементов в строке, в которой они находятся.
rows, cols = my_arr_2.shape
for i in range(rows):
    otr = np.sum(my_arr_2[i] < 0)
    my_arr_2[i][my_arr_2[i] < 0] = otr
print(my_arr_2)

[[3 1 2 2 1]
 [5 1 3 5 1]
 [2 0 1 2 5]
 [3 2 4 0 1]
 [2 0 1 4 2]]


# Pandas

In [14]:
df = pd.read_csv("https://raw.githubusercontent.com/rolandmueller/titanic/main/titanic3.csv")

**Задание 1**
+ Какой была максимальная цена билета? Медиана?
+ Какова доля выживших?
+ Сколько мужчин / женщин находилось на борту?

<b><span style='color:Green'>Ответ:
- <span style='color:teal'> Максимальная цена билета составила 512.3292, а медианная цена 14.4542
- <span style='color:teal'> Доля выживших: 38.2%
- <span style='color:teal'> На борту находилось 843 мужчины и 466 женщин

In [15]:
df.head(3)

pclass  survived                            name     sex      age  sibsp  \
0       1         1   Allen, Miss. Elisabeth Walton  female  29.0000      0   
1       1         1  Allison, Master. Hudson Trevor    male   0.9167      1   
2       1         0    Allison, Miss. Helen Loraine  female   2.0000      1   

   parch  ticket      fare    cabin embarked boat  body  \
0      0   24160  211.3375       B5        S    2   NaN   
1      2  113781  151.5500  C22 C26        S   11   NaN   
2      2  113781  151.5500  C22 C26        S  NaN   NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


In [17]:
df.isna().sum()

pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64

In [18]:
# Проверим на дубликаты
df.duplicated().sum()

0

In [19]:
# Какой была максимальная цена билета? 
df['fare'].max()

512.3292

In [20]:
# Какой была медиана?
df['fare'].median()

14.4542

In [21]:
c = ((df['survived'].sum())/(df['survived'].count())*100).round(2)
print(f"Доля выживших составила {c}%")

Доля выживших составила 38.2%


In [26]:
# Сколько мужчин / женщин находилось на борту?
df['sex'].value_counts()

male      843
female    466
Name: sex, dtype: int64

**Задание 2**
+ Выведите распределение переменной __pclass__ для мужчин и женщин.
    + Сколько мужчин было в первом классе?
    + Сколько женщин было в третьем классе?
      
+ Создайте новый признак __age_group__, который равен:
    + 1, если пассажир был моложе 30 лет;
    + 2, если пассажиру было от 30 до 60 лет;
    + 3, если пассажир был старше 60 лет.

<b><span style='color:Green'>Ответ:
- <span style='color:teal'>В первом классе было 179 мужчин
- <span style='color:teal'>В третьем классе было 216 женщин

In [27]:
# Выведите распределение переменной pclass для мужчин и женщин.
l = ['male','female']
l_1 = [1,2,3]
for s in l:
    for p in l_1:
        a = df['pclass'][((df['pclass'] == p) & (df['sex'] == s))].count()
        print(f"В {p} классе было {a} {s}")

В 1 классе было 179 male
В 2 классе было 171 male
В 3 классе было 493 male
В 1 классе было 144 female
В 2 классе было 106 female
В 3 классе было 216 female


In [28]:
# Создайте новый признак age_group, который равен:
smth = []
for i in df['age']:
    if i<30:
        smth.append(1)
    elif 30<= i <=60:
        smth.append(2)
    else:
        smth.append(3)
df['age_group']=smth

**Задание 3**\
Найдите самое популярное имя среди пассажиров:
+ мужчин;
+ женщин.

<b><span style='color:Green'>Ответ:
- <span style='color:teal'>У мужчин самое популярное имя - William
- <span style='color:teal'>У женщин самое популярное имя - Mary

In [29]:
#Самое популярное мужское имя
first_name_men = df.loc[df['sex'] == 'male', 'name'].apply(
    lambda name_men: name_men.split(',')[1].split()[1]
)
first_name_men.value_counts().head(1)

William    50
Name: name, dtype: int64

In [30]:
# Посмотрим как записывались женские имена
df.loc[df['sex']=='female','name']

0                         Allen, Miss. Elisabeth Walton
2                          Allison, Miss. Helen Loraine
4       Allison, Mrs. Hudson J C (Bessie Waldo Daniels)
6                     Andrews, Miss. Kornelia Theodosia
8         Appleton, Mrs. Edward Dale (Charlotte Lamson)
                             ...                       
1286    Whabee, Mrs. George Joseph (Shawneene Abi-Saab)
1290                   Wilkes, Mrs. James (Ellen Needs)
1300            Yasbeck, Mrs. Antoni (Selini Alexander)
1304                               Zabour, Miss. Hileni
1305                              Zabour, Miss. Thamine
Name: name, Length: 466, dtype: object

In [31]:
def first_name_female(name):
    first_w = name.str.extract(r"Mrs\.\s+[^(]*\((\w+)", expand=False)
    first_w.loc[first_w.isna()] = name.str.extract(r"\.\s+(\w+)", expand=False)
    return first_w

In [32]:
# Самое популярное женское имя
first_name_women = first_name_female(df.loc[df['sex']=='female','name'])
first_name_women.value_counts().head(1)

Mary    22
Name: name, dtype: int64

**Задание 4**
+ Выведите данные о всех пассажирах, которые сели в Cherbourg (Embarked='C') и были старше 50 лет.
+ Выведите данные о всех женщинах младше 20 и старше 50 лет.
+ Сколько пассажиров старше 70 лет было на борту?

<b><span style='color:Green'>Ответ:
- <span style='color:teal'> Всего 27 пассажиров, которые сели в Cherbourg (Embarked='C') и были старше 50 лет.
- <span style='color:teal'> 135 женщин на борту были либо младше 20, либо старше 50.
- <span style='color:teal'> На борту было 6 пассажиров старше 70 лет    

In [33]:
# Выведите данные о всех пассажирах, которые сели в Cherbourg (Embarked='C') и были старше 50 лет.
df[(df['embarked']=='C')&(df['age']>50.0)]

pclass  survived                                               name  \
9         1         0                            Artagaveytia, Mr. Ramon   
43        1         1    Bucknell, Mrs. William Robert (Emma Eliza Ward)   
48        1         1   Candee, Mrs. Edward (Helen Churchill Hungerford)   
50        1         1  Cardeza, Mrs. James Warburton Martinez (Charlo...   
78        1         1  Compton, Mrs. Alexander Taylor (Mary Eliza Ing...   
104       1         1                     Eustis, Miss. Elizabeth Mussey   
123       1         1                   Frolicher-Stehli, Mr. Maxmillian   
135       1         0                          Goldschmidt, Mr. George B   
136       1         1                          Gracie, Col. Archibald IV   
174       1         0                            Kent, Mr. Edward Austin   
186       1         1          Lindstrom, Mrs. Carl Johan (Sigrid Posse)   
192       1         1                               Lurette, Miss. Elise   
215       1         0                         Newell, Mr. Arthur Webster   
221       1         0                     Ostby, Mr. Engelhart Cornelius   
233       1         1      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)   
246       1         0                             Rothschild, Mr. Martin   
247       1         1     Rothschild, Mrs. Martin (Elizabeth L. Barrett)   
252       1         0                         Ryerson, Mr. Arthur Larned   
264       1         1                Simonius-Blumer, Col. Oberst Alfons   
267       1         0                            Smith, Mr. James Clinch   
276       1         0                      Spencer, Mr. William Augustus   
280       1         1                    Stengel, Mr. Charles Emil Henry   
282       1         1    Stephenson, Mrs. Walter Bertram (Martha Eustis)   
303       1         0                           Warren, Mr. Frank Manley   
304       1         1   Warren, Mrs. Frank Manley (Anna Sophia Atkinson)   
308       1         1              White, Mrs. John Stuart (Ella Holmes)   
316       1         0                        Williams, Mr. Charles Duane   

        sex   age  sibsp  parch    ticket      fare            cabin embarked  \
9      male  71.0      0      0  PC 17609   49.5042              NaN        C   
43   female  60.0      0      0     11813   76.2917              D15        C   
48   female  53.0      0      0  PC 17606   27.4458              NaN        C   
50   female  58.0      0      1  PC 17755  512.3292      B51 B53 B55        C   
78   female  64.0      0      2  PC 17756   83.1583              E45        C   
104  female  54.0      1      0     36947   78.2667              D20        C   
123    male  60.0      1      1     13567   79.2000              B41        C   
135    male  71.0      0      0  PC 17754   34.6542               A5        C   
136    male  53.0      0      0    113780   28.5000              C51        C   
174    male  58.0      0      0     11771   29.7000              B37        C   
186  female  55.0      0      0    112377   27.7208              NaN        C   
192  female  58.0      0      0  PC 17569  146.5208              B80        C   
215    male  58.0      0      2     35273  113.2750              D48        C   
221    male  65.0      0      1    113509   61.9792              B30        C   
233  female  56.0      0      1     11767   83.1583              C50        C   
246    male  55.0      1      0  PC 17603   59.4000              NaN        C   
247  female  54.0      1      0  PC 17603   59.4000              NaN        C   
252    male  61.0      1      3  PC 17608  262.3750  B57 B59 B63 B66        C   
264    male  56.0      0      0     13213   35.5000              A26        C   
267    male  56.0      0      0     17764   30.6958               A7        C   
276    male  57.0      1      0  PC 17569  146.5208              B78        C   
280    male  54.0      1      0     11778   55.4417             C116        C   
282  female  

In [34]:
# Выведите данные о всех женщинах младше 20 и старше 50 лет.
df.loc[(df['sex']=='female')&((df['age']<20.0)|(df['age']>50.0))]

pclass  survived                                               name  \
2          1         0                       Allison, Miss. Helen Loraine   
6          1         1                  Andrews, Miss. Kornelia Theodosia   
8          1         1      Appleton, Mrs. Edward Dale (Charlotte Lamson)   
11         1         1  Astor, Mrs. John Jacob (Madeleine Talmadge Force)   
27         1         1            Bishop, Mrs. Dickinson H (Helen Walton)   
...      ...       ...                                                ...   
1265       3         0                          Van Impe, Miss. Catharina   
1273       3         0                 Vander Planke, Miss. Augusta Maria   
1279       3         0               Vestrom, Miss. Hulda Amanda Adolfina   
1300       3         1            Yasbeck, Mrs. Antoni (Selini Alexander)   
1304       3         0                               Zabour, Miss. Hileni   

         sex   age  sibsp  parch    ticket      fare    cabin embarked boat  \
2     female   2.0      1      2    113781  151.5500  C22 C26        S  NaN   
6     female  63.0      1      0     13502   77.9583       D7        S   10   
8     female  53.0      2      0     11769   51.4792     C101        S    D   
11    female  18.0      1      0  PC 17757  227.5250  C62 C64        C    4   
27    female  19.0      1      0     11967   91.0792      B49        C    7   
...      ...   ...    ...    ...       ...       ...      ...      ...  ...   
1265  female  10.0      0      2    345773   24.1500      NaN        S  NaN   
1273  female  18.0      2      0    345764   18.0000      NaN        S  NaN   
1279  female  14.0      0      0    350406    7.8542      NaN        S  NaN   
1300  female  15.0      1      0      2659   14.4542      NaN        C  NaN   
1304  female  14.5      1      0      2665   14.4542      NaN        C  NaN   

       body                        home.dest  age_group  
2       NaN  Montreal, PQ / Chesterville, ON          1  
6       NaN                       Hudson, NY          3  
8       NaN              Bayside, Queens, NY          2  
11      NaN                     New York, NY          1  
27      NaN                     Dowagiac, MI          1  
...     ...                              ...        ...  
1265    NaN                              NaN          1  
1273    NaN                              NaN          1  
1279    NaN                              NaN          1  
1300    NaN                              NaN          1  
1304  328.0                              NaN          1  

[135 rows x 15 columns]

In [35]:
# Сколько пассажиров старше 70 лет было на борту?
df['name'][df['age']>70.0].count()

6

**Задание 5**
+ Посчитайте средний возраст среди мужчин и среди женщин.
+ Посчитайте доли выживших в различных возрастных группах (новый признак **AgeGroup**).
+ Найдите максимальную стоимость билета в каждом из ценовых классов.

<b><span style='color:Green'>Ответ:
- <span style='color:teal'> Средний возраст мужчин сочтавил 30.59 лет, а возраст женщин - 28.69 лет
- <span style='color:teal'> Доля выживших в первой возрастной группе (до 30 лет ) составила 0.41; во второй группе (от 30 до 60 лет включительно) - 0.42, а в третьей (старше 60 лет) - 0.27
- <span style='color:teal'> Максимальная стоимость билетов в первом классе достигла - 512.33, во втором классе - 73.50, в третьем - 69.55

In [36]:
# Посчитайте средний возраст среди мужчин и среди женщин.
mean_age=df.pivot_table(index = 'sex',
                       aggfunc = 'mean',
                       values = 'age').round(2)
mean_age

age
sex          
female  28.69
male    30.59

In [37]:
pd.crosstab(df['sex'],'average age',values = df['age'],aggfunc = 'mean').round(2)

col_0   average age
sex                
female        28.69
male          30.59

In [38]:
# Посчитайте доли выживших в различных возрастных группах (новый признак AgeGroup).
dolya_surv = df.pivot_table(index = 'age_group',
                           aggfunc = 'mean',
                           values = 'survived').round(2)
dolya_surv.columns=['Доля выживших']
dolya_surv

Доля выживших
age_group               
1                   0.41
2                   0.42
3                   0.27

In [39]:
pd.crosstab(df['age_group'],'survival rate',values = df['survived'],aggfunc = 'mean').round(2)

col_0      survival rate
age_group               
1                   0.41
2                   0.42
3                   0.27

In [40]:
# Найдите максимальную стоимость билета в каждом из ценовых классов.
pclass_max = df.pivot_table(index = 'pclass',
                           aggfunc = 'max',
                           values = 'fare').round(2)
pclass_max.columns = ['Макс. стоимость билета']
pclass_max

Макс. стоимость билета
pclass                        
1                       512.33
2                        73.50
3                        69.55

In [41]:
pd.crosstab(df['pclass'],'max_fare',values = df['fare'],aggfunc = 'max').round(2)

col_0   max_fare
pclass          
1         512.33
2          73.50
3          69.55

**Задание 6**\
Ответьте на предыдущие вопросы с помощью функции crosstab и метода pivot_table.